certbot/compat/filesystem.py

```diff
@@ -36,6 +43,67 @@ def chmod(file_path, mode):
         _apply_win_mode(file_path, mode)
 
 
+# Funtion os.makedirs will temporarily monkeypatch os.mkdir to replace it by filesystem.mkdir.
+# However, filesystem.mkdir also invokes os.mkdir. To avoid a looped reference, we take now a
+# reference of the original os.mkdir, to use it in filesystem.mkdir.
+_os_mkdir = os.mkdir
+
+
+def makedirs(file_path, mode=0o777, exists_ok=False):
+    # type: (str, int, bool) -> None
+    """
+    Rewrite of original os.makedirs function, that will ensure on Windows that given mode
+    is correctly applied.
+    :param str file_path: The file path to open
+    :param int mode: POSIX mode to apply on leaf directory when created, Python defaults
+                     will be applied if ``None``
+    :param bool exists_ok: If set to ``True``, do not raise exception if leaf directory
+                           already exists.
+    """
+    # As we know that os.mkdir is called internally by os.makedirs, we will swap the function in
+    # os module for the time of makedirs execution.
+    orig_mkdir_fn = os.mkdir
+    try:
+        os.mkdir = mkdir  # type: ignore
+        try:
+            os.makedirs(file_path, mode)
+        except (IOError, OSError) as err:
+            # In case of exists_ok is True, and exception was about the path to already exist,
+            # we ignore the exception.
+            if not exists_ok or err.errno != errno.EEXIST:
+                raise
+    finally:
+        os.mkdir = orig_mkdir_fn
+
+
+def mkdir(file_path, mode=0o777):
+    # type: (str, int) -> None
+    """
+    Rewrite of original os.mkdir function, that will ensure on Windows that given mode
+    is correctly applied.
+    :param str file_path: The file path to open
+    :param int mode: POSIX mode to apply on directory when created, Python defaults
+                     will be applied if ``None``
+    """
+    if POSIX_MODE:
+        _os_mkdir(file_path, mode)
+    else:
+        attributes = win32security.SECURITY_ATTRIBUTES()
+        security = attributes.SECURITY_DESCRIPTOR
+        user = _get_current_user()
+        dacl = _generate_dacl(user, mode)
+        security.SetSecurityDescriptorDacl(1, dacl, 0)
+
+        try:
+            win32file.CreateDirectory(file_path, attributes)
+        except pywintypes.error as err:
+            # Handle native windows error into python error to be consistent with the API
+            # of os.mkdir in the situation of a directory already existing.
+            if err.winerror == winerror.ERROR_ALREADY_EXISTS:
+                raise OSError(errno.EEXIST, err.strerror)
```

Comment:
sure do wish we could drop 2.7 and raise a py3 `FileExistsError` rather than `OSError`... probably best to stick with this though, I suppose.

Reply:
I would be tempted to use `FileExistsError` since it subclasses correctly `OSError`, and that this portion of code should run only on Python 3 so we can suppose that the error subclass exists at this point. However I agree with you that it is less obvious when reading the code, so it seems that current implementation is shorter and cleaner for now.

Maybe a refactoring to do once we drop completely Python 2.7.

Reply:
as I mentioned in the conversation above, since this code will only ever be run on py3, I'm excited about moving to py3-style as much as possible now.

Reply:
See general comment below.